plugins/ltac/tacsubst.ml

```diff
@@ -172,54 +172,57 @@ let rec subst_atomic subst (t:glob_atomic_tactic_expr) = match t with
       TacInversion (InversionUsing (subst_glob_constr subst c,cl),hyp)
 
 and subst_tactic subst (t:glob_tactic_expr) = match t with
-  | TacAtom { CAst.v=t } -> TacAtom (CAst.make @@ subst_atomic subst t)
-  | TacFun tacfun -> TacFun (subst_tactic_fun subst tacfun)
-  | TacLetIn (r,l,u) ->
+  | TacAtom { CAst.loc; v=t } -> TacAtom (CAst.make ?loc @@ subst_atomic subst t)
+  | TacFun { CAst.loc; v=tacfun } -> TacFun (CAst.make ?loc(subst_tactic_fun subst tacfun))
+  | TacLetIn { CAst.loc; v=(r,l,u) } ->
       let l = List.map (fun (n,b) -> (n,subst_tacarg subst b)) l in
-      TacLetIn (r,l,subst_tactic subst u)
-  | TacMatchGoal (lz,lr,lmr) ->
-      TacMatchGoal(lz,lr, subst_match_rule subst lmr)
-  | TacMatch (lz,c,lmr) ->
-      TacMatch (lz,subst_tactic subst c,subst_match_rule subst lmr)
-  | TacId _ | TacFail _ as x -> x
-  | TacProgress tac -> TacProgress (subst_tactic subst tac:glob_tactic_expr)
-  | TacShowHyps tac -> TacShowHyps (subst_tactic subst tac:glob_tactic_expr)
-  | TacAbstract (tac,s) -> TacAbstract (subst_tactic subst tac,s)
-  | TacThen (t1,t2) ->
-      TacThen (subst_tactic subst t1, subst_tactic subst t2)
-  | TacDispatch tl -> TacDispatch (List.map (subst_tactic subst) tl)
-  | TacExtendTac (tf,t,tl) ->
-      TacExtendTac (Array.map (subst_tactic subst) tf,
+      TacLetIn (CAst.make ?loc (r,l,subst_tactic subst u))
+  | TacMatchGoal { CAst.loc; v=(lz,lr,lmr) } ->
+      TacMatchGoal (CAst.make ?loc (lz,lr, subst_match_rule subst lmr))
+  | TacMatch { CAst.loc; v=(lz,c,lmr) } ->
+      TacMatch (CAst.make ?loc (lz,subst_tactic subst c,subst_match_rule subst lmr))
+  | TacId {CAst.loc; v=t} as x -> x
+  | TacFail  {CAst.loc; v=t} as x -> x
+  | TacProgress { CAst.loc; v=tac } -> TacProgress (CAst.make ?loc (subst_tactic subst tac:glob_tactic_expr))
```

Comment:
There are combinators in CAst that help with this pattern.

Reply:
Can you give me a concrete example?  Are you referring to the "{ CAst.loc; v=tac }" or the "(CAst.make ?loc "?  Converting this stuff is really tedious.

Reply:
grep for uses of `CASt.map` and `CAst.with` [really a fold]

If you are careful when structuring your tactic ast mappers you can just wrap them properly, in the usual way for fold / map transformations.